server/src/test/java/org/elasticsearch/persistent/PersistentTasksNodeServiceTests.java

```diff
@@ -303,6 +305,70 @@ public void sendCompletionRequest(final String taskId, final long taskAllocation
+                                              final Exception taskFailure, final String localAbortReason,
+                                              final ActionListener<PersistentTask<?>> listener) {
+                capturedTaskId.set(taskId);
+                capturedListener.set(listener);
+                capturedLocalAbortReason.set(localAbortReason);
+                assertThat(taskFailure, nullValue());
+            }
+        };
+        @SuppressWarnings("unchecked") PersistentTasksExecutor<TestParams> action = mock(PersistentTasksExecutor.class);
+        when(action.getExecutor()).thenReturn(ThreadPool.Names.SAME);
+        when(action.getTaskName()).thenReturn("test");
+        when(action.createTask(anyLong(), anyString(), anyString(), any(), any(), any()))
+            .thenReturn(new TestPersistentTasksPlugin.TestTask(1, "persistent", "test", "", new TaskId("cluster", 1),
+                Collections.emptyMap()));
+        PersistentTasksExecutorRegistry registry = new PersistentTasksExecutorRegistry(Collections.singletonList(action));
+
+        int nonLocalNodesCount = randomInt(10);
+        MockExecutor executor = new MockExecutor();
+        TaskManager taskManager = new TaskManager(Settings.EMPTY, threadPool, Collections.emptySet());
+        PersistentTasksNodeService coordinator = new PersistentTasksNodeService(persistentTasksService,
+            registry, taskManager, executor);
+
+        ClusterState state = createInitialClusterState(nonLocalNodesCount, Settings.EMPTY);
+
+        // Allocate first task
+        ClusterState newClusterState = addTask(state, "test", null, "this_node");
+        coordinator.clusterChanged(new ClusterChangedEvent("test", newClusterState, state));
+
+        // Check the task is known to the task manager
+        assertThat(taskManager.getTasks().size(), equalTo(1));
+        AllocatedPersistentTask runningTask = (AllocatedPersistentTask) taskManager.getTasks().values().iterator().next();
+        String persistentId = runningTask.getPersistentTaskId();
+        // Make sure it returns correct status
+        Task.Status status = runningTask.getStatus();
+        assertThat(status.toString(), equalTo("{\"state\":\"STARTED\"}"));
+
+        // Abort the task locally
+        runningTask.markAsLocallyAborted("testing local abort");
+
+        // That should trigger an unassignment request
+        assertThat(capturedTaskId.get(), equalTo(persistentId));
+        assertThat(capturedLocalAbortReason.get(), equalTo("testing local abort"));
+        // Notify successful unassignment
+        PersistentTasksCustomMetadata persistentTasksMetadata = newClusterState.getMetadata().custom(PersistentTasksCustomMetadata.TYPE);
+        capturedListener.get().onResponse(persistentTasksMetadata.getTask(persistentId));
+
+        // Check the task is now removed from the local task manager
+        assertThat(taskManager.getTasks().values(), empty());
+    }
```

Comment:
Can we try to randomly either abort, complete or fail the task afterwards and check that it does nothing/throws as per my comment above?